pkg/ingester/ingester.go

```diff
@@ -760,6 +760,11 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) (
 
 	db, err := i.getOrCreateTSDB(userID, false)
 	if err != nil {
+		// Check for a particular per-instance limit and return that error directly
+		// since it contains extra information for gRPC and our logging middleware.
+		if errors.Is(err, errMaxTenantsReached) {
```

Comment:
So why do we need this logic at all?

Reply:
Could you improve the existing tests to assert on the returned error and check that gRPC `status.FromError()` returns the expected code?

Reply:
> So why do we need this logic at all?

Because `wrapWithUser()` as it exists today doesn't actually "wrap" the error. It does something like `errors.New(err.Error())` and so it returns a brand new error that doesn't have any relation to the original error or implement the `GRPCStatus()` or `ShouldLog()` methods. I could change it but it seemed like it purposefully didn't wrap the existing error.

> Could you improve the existing tests to assert on the returned error and check that gRPC `status.FromError()` returns the expected code?

Sure, will do.

Reply:
> Because wrapWithUser() as it exists today doesn't actually "wrap" the error.

Ouf... you're right, again.

> I could change it but it seemed like it purposefully didn't wrap the existing error.

No, don't do it. I thin would be unsafe. There's some context here https://github.com/cortexproject/cortex/pull/2004 but TL;DR is that some errors carry the series labels but these series labels are only safe to read during the execution of `push()` because they're unmarshalled from the write request into a pool. So before returning, we effectively "make a copy" to ensure the error will be safe after the `push()` has returned.

Reply:
The `errMaxTenantsReached` and `errMaxInMemorySeriesReached` don't carry any label, so they're expected to be safe.

As a retrospective, we could have better handled the returned error, maybe implementing a specific interface for unsafe error messages, because it's currently very obscure. But that's for another day :)

Reply:
> As a retrospective, we could have better handled the returned error, maybe implementing a specific interface for unsafe error messages, because it's currently very obscure. But that's for another day :)

IMO we should create an issue and tackle that, as otherwise this is a recipe for a disaster, and it's just a matter of time that it explodes somewhere.

Errors are something that is meant to escape your function, we shouldn't return errors that aren't safe to escape further than _some point_.